### PR TITLE
Optimize riak to s3 migration

### PR DESCRIPTION
**Pull request by millerdev**
Use gevent to migrate multiple blobs simultaneously. The first four commits are (p)refactoring in preparation for ~the final commit~ 5368d38b7b6c1c5ed2344e6b99ab6155498f266a, which adds parallel processing.

Review by commit 🐠 

@nickpell @jmtroth0

### CODE REVIEW

**Review by nickpell — APPROVED** [review]
yay parallelism, yay refactoring

File: corehq/blobs/migrate.py [hunk 1]
```diff
@@ -365,17 +387,89 @@ def get_document_provider(self):
...
+        retry_count = retry_blobs.setdefault(key, 0)
+        # pool.spawn will block until a worker is available
+        pool.spawn(work_on, doc, key, retry_count)
+        # Returning True here means the underlying iterator will think
+        # this doc has been processed successfully. Therefore we must
+        # process this doc before the process exits or save it to be
+        # processed on the next run.
+        return True
+
+    queue = LifoQueue()
```
Comment (jmtroth0) [inline corehq/blobs/migrate.py]: Why `lifo`?

**Review by jmtroth0 — COMMENTED** [review]
just a couple things

File: corehq/blobs/management/commands/run_blob_migration.py [hunk 2]
```diff
@@ -45,19 +45,33 @@ def add_arguments(self, parser):
...
-    def handle(self, slug=None, log_dir=None, reset=False, chunk_size=100,
-               **options):
+    def handle(self, slug, log_dir=None, **options):
         try:
             migrator = MIGRATIONS[slug]
         except KeyError:
             raise CommandError(USAGE)
+        if not migrator.has_worker_pool:
+            num_workers = options.pop("num_workers")
+            if num_workers != DEFAULT_WORKER_POOL_SIZE:
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/blobs/management/commands/run_blob_migration.py]: The warning is only displayed if the value is different from the default, in which case this code is inferring that the user explicitly specified a value when running the command. One annoying edge case is if they specify a value _and_ it happens to be the default they will not see this warning. I don't think that's a big enough concern to be worth pursuing.

File: corehq/blobs/migrate.py (hunk 1 above)
Comment (millerdev) [reply corehq/blobs/migrate.py]: So the retry loop does not process all items in the queue before stopping (on keyboard interrupt, etc.). Another option is to use a different signal/event channel and block on the queue with a timeout—checking for the stop event periodically. More complicated, probably doesn't matter that much.

File: corehq/blobs/management/commands/run_blob_migration.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/blobs/management/commands/run_blob_migration.py]: Oooh now i'm understanding. I definitely misread `not migrator.has_worker_pool`

**Pull request merged**